Fix invalid float literal suggestions when recovering an integer

cassaundra opened the pull request:
Only suggest adding a zero to integers with a preceding dot when the change will result in a valid floating point literal.

For example, `.0x0` should not be turned into `0.0x0`.

r? nnethercote

cassaundra commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -1840,7 +1840,11 @@ impl<'a> Parser<'a> {
-                    if self.token.span.hi() == next_token.span.lo() {
+                    // If this integer looks like a float, then recover as such.
+                    if suffix.map_or(true, |s| s == sym::f32 || s == sym::f64)
+                        && symbol.as_str().chars().all(|c| c.is_numeric() || c == '_')
```
Something that might deserve a comment: we will never encounter the exponent part of a floating point literal here, since the characters (`e` or `E`, and sometimes `+` or `-`) do not overlap with integer literals, so we only need to check that all characters are numeric.

nnethercote replied:
Everything from "we will" onwards would make a great comment here.

cassaundra commented on src/test/ui/suggestions/recover-invalid-float-invalid.rs:
```diff
@@ -0,0 +1,21 @@
+fn main() {}
+
+fn a() {
```
Not sure if there's a better way to test errors like this other than putting them in separate functions/items.

nnethercote replied:
They can all go in a single function.

cassaundra replied:
Unfortunately it seems having two of these errors in a single function means only the first will be tested, which is why I had them in separate functions. Just curious if there's a more idiomatic way to do this, but I'm not seeing anything. Not blocking.

nnethercote replied:
How are you writing the test? There are lots of examples in the test suite with more than one error per function, e.g. src/test/ui/attrs-resolution-errors.rs. Also see https://rustc-dev-guide.rust-lang.org/tests/ui.html for docs about writing UI tests.

cassaundra replied:
Something like the following fails prematurely:
```rust
fn main() {
    _ = .3u32;
    //~^ ERROR expected expression, found `.`
    _ = .0b0;
    //~^ ERROR expected expression, found `.`
    _ = .0o07;
    //~^ ERROR expected expression, found `.`
    _ = .0x0ABC;
    //~^ ERROR expected expression, found `.`
}
```

nnethercote replied:
Oh, I see, it's nothing to do with the test harness, it's the parser itself no longer able to recover. Hmm.

@compiler-errors: This PR is about some error suggestions relating to invalid literals. E.g. if you have `.3u32` the compiler current suggests:
```
error: float literals must have an integer part
 --> c.rs:2:9
  |
2 |     _ = .3u32;
  |         ^^^^^ help: must have an integer part: `0.3u32`
```
But that suggestion is bogus, because `0.3u32` isn't a valid literal:
```
error: invalid suffix `u32` for float literal
 --> c.rs:2:9
  |
2 |     _ = 0.3u32;
  |         ^^^^^^ invalid suffix `u32`
  |
  = help: valid suffixes are `f32` and `f64`
```
So I suggested that @cassaundra fix this, and this PR avoids the suggestion if the suffix isn't `f32`/`f64`, changing the error  to:
```
error: expected expression, found `.`
 --> c.rs:2:9
  |
2 |     _ = .3u32;
  |         ^ expected expression
```
This also prevents error recovery from happening, so any subsequent parse errors in the function aren't picked up.

Do you think this change is good enough as is? `.3u32` is weird enough that I'm not sure if trying to recover makes sense, but I'm no expert on error messages.

estebank replied:
@nnethercote yes, failing the parse and consuming the rest of the block is unfortunate but acceptable. We can file a ticket to follow up on this and attempt to recover by discarding only until the `;` (which I thought we did already in statements, but there might be some reason it doesn't trigger here). Splitting the test in one function per unparseable case should be enough to get around this issue in the tests.

nnethercote commented on src/test/ui/suggestions/recover-invalid-float-invalid.rs:
```diff
@@ -0,0 +1,21 @@
+fn main() {}
```
A brief comment at the top of the file describing what's being tested would be good.

nnethercote reviewed (commented):
Looking good, just some minor suggestions.

cassaundra commented:
Discussion aside, requested comments have been added.

pnkfelix commented:
@bors r+

(I think the test is fine. If we end up improving parser recovery to enable such tests to be more concise, that would be great, but I don't think it makes sense for that to block landing this PR.)

bors commented:
:pushpin: Commit 7497f7c5ac370ec1c9771b0d39f39f47d6d0ee32 has been approved by `pnkfelix`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pnkfelix 7497f7c5ac370ec1c9771b0d39f39f47d6d0ee32 -->
<!-- homu: {"type":"Approved","sha":"7497f7c5ac370ec1c9771b0d39f39f47d6d0ee32","approver":"pnkfelix","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
tests moved from `src/test` to `tests`

nnethercote commented:
@cassaundra This just needs to be updated for the recent `src/test/*`-to-`tests/*` renaming. Then it will be ready to merge.

cassaundra committed:
Fix invalid float literal suggestions when recovering an integer

Only suggest adding a zero to integers with a preceding dot when the change will
result in a valid floating point literal.

For example, `.0x0` should not be turned into `0.0x0`.

cassaundra commented:
Ah apologies, I missed that. Should be fixed now.

cassaundra commented:
@rustbot ready

estebank commented:
@bors r=pnkfelix

bors commented:
:pushpin: Commit 80fcd7c40ec46613b130b5d732162ecbfda6c509 has been approved by `pnkfelix`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pnkfelix 80fcd7c40ec46613b130b5d732162ecbfda6c509 -->
<!-- homu: {"type":"Approved","sha":"80fcd7c40ec46613b130b5d732162ecbfda6c509","approver":"pnkfelix","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 80fcd7c40ec46613b130b5d732162ecbfda6c509 with merge 487e83b711b96d84ce2edfc4cb4ba792c10b589e...
<!-- homu: {"type":"BuildStarted","head_sha":"80fcd7c40ec46613b130b5d732162ecbfda6c509","merge_sha":"487e83b711b96d84ce2edfc4cb4ba792c10b589e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4049323545/jobs/6967433773)
Approved by: pnkfelix
Pushing 487e83b711b96d84ce2edfc4cb4ba792c10b589e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4049323545/jobs/6967433773"},"merge_sha":"487e83b711b96d84ce2edfc4cb4ba792c10b589e"} -->

rust-timer commented:
Finished benchmarking commit (487e83b711b96d84ce2edfc4cb4ba792c10b589e): [comparison URL](https://perf.rust-lang.org/compare.html?start=001a77fac33f6560ff361ff38f661ff5f1c6bf85&end=487e83b711b96d84ce2edfc4cb4ba792c10b589e&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=001a77fac33f6560ff361ff38f661ff5f1c6bf85&end=487e83b711b96d84ce2edfc4cb4ba792c10b589e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.4% | [-5.4%, -5.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=001a77fac33f6560ff361ff38f661ff5f1c6bf85&end=487e83b711b96d84ce2edfc4cb4ba792c10b589e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.8% | [-3.8%, -3.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-4.3%, -1.0%] | 2     |
| All ❌✅ (primary)                 | -3.8% | [-3.8%, -3.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=001a77fac33f6560ff361ff38f661ff5f1c6bf85&end=487e83b711b96d84ce2edfc4cb4ba792c10b589e&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.4% | [-3.4%, -3.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
